snow/validators/manager.go

```diff
@@ -49,27 +109,187 @@ type manager struct {
+	// of this function to ensure that errors are returned consistently.
+	//
+	// Consider the case that:
+	//	AddStaker(subnetID, nodeID, 1)
+	//	go func() {
+	//		AddWeight(subnetID, nodeID, 1)
+	//	}
+	//	go func() {
+	//		RemoveWeight(subnetID, nodeID, 1)
+	//	}
+	//
+	// In this case, after both goroutines have finished, either AddWeight
+	// should have errored, or the weight of the node should equal 1. It would
+	// be unexpected to not have received an error from AddWeight but for the
+	// node to no longer be tracked as a validator.
 	m.lock.RLock()
 	defer m.lock.RUnlock()
 
-	vdrs, ok := m.subnetToVdrs[subnetID]
-	return vdrs, ok
+	set, exists := m.subnetToVdrs[subnetID]
+	if !exists {
+		return errMissingValidator
+	}
+
+	return set.AddWeight(nodeID, weight)
+}
+
+func (m *manager) GetWeight(subnetID ids.ID, nodeID ids.NodeID) uint64 {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return 0
+	}
+
+	return set.GetWeight(nodeID)
+}
+
+func (m *manager) GetValidator(subnetID ids.ID, nodeID ids.NodeID) (*Validator, bool) {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return nil, false
+	}
+
+	return set.Get(nodeID)
+}
+
+func (m *manager) SubsetWeight(subnetID ids.ID, validatorIDs set.Set[ids.NodeID]) (uint64, error) {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return 0, nil
+	}
+
+	return set.SubsetWeight(validatorIDs)
+}
+
+func (m *manager) RemoveWeight(subnetID ids.ID, nodeID ids.NodeID, weight uint64) error {
+	if weight == 0 {
+		return ErrZeroWeight
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	set, exists := m.subnetToVdrs[subnetID]
+	if !exists {
+		return errMissingValidator
+	}
+
+	if err := set.RemoveWeight(nodeID, weight); err != nil {
+		return err
+	}
+	// If this was the last validator in the subnet and no callback listeners
+	// are registered, remove the subnet
+	if set.Len() == 0 && !set.HasCallbackRegistered() {
+		delete(m.subnetToVdrs, subnetID)
+	}
+
+	return nil
+}
+
+func (m *manager) Contains(subnetID ids.ID, nodeID ids.NodeID) bool {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return false
+	}
+
+	return set.Contains(nodeID)
+}
+
+func (m *manager) Count(subnetID ids.ID) int {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return 0
+	}
+
+	return set.Len()
+}
+
+func (m *manager) TotalWeight(subnetID ids.ID) (uint64, error) {
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return 0, nil
+	}
+
+	return set.TotalWeight()
+}
+
+func (m *manager) Sample(subnetID ids.ID, size int) ([]ids.NodeID, error) {
+	if size == 0 {
+		return nil, nil
+	}
+
+	m.lock.RLock()
+	set, exists := m.subnetToVdrs[subnetID]
+	m.lock.RUnlock()
+	if !exists {
+		return nil, ErrMissingValidators
```

Comment:
We have inconsistent behavior on this API because some functions return an error vs a zero value if we aren't tracking `subnetID`. Would it make sense for us to return a bool/error if the subnet isn't found for all apis?

Reply:
This isn't really inconsistent - the difference here is that the returned `[]ids.NodeID` slice is expected to always have length `size`. If the total weight of the validator set is less than `size` an error is also returned.